contracts/peg/VanillaPriceBoundPSM.sol

```diff
@@ -0,0 +1,131 @@
+
+    /// @notice constructor
+    /// @param _floor minimum acceptable oracle price
+    /// @param _ceiling maximum  acceptable oracle price
+    /// @param _params PSM construction params
+    constructor(
+        uint256 _floor,
+        uint256 _ceiling,
+        OracleParams memory _params,
+        uint256 _reservesThreshold,
+        IERC20 _underlyingToken,
+        IPCVDeposit _surplusTarget
+    )
+        VanillaPSM(
+            _params,
+            _reservesThreshold,
+            _underlyingToken,
+            _surplusTarget
+        )
+    {
+        _setCeilingBasisPoints(_ceiling);
+        _setFloorBasisPoints(_floor);
+    }
+
+    /// @notice sets the floor price in BP
+    function setOracleFloorBasisPoints(uint256 newFloorBasisPoints)
+        external
+        override
+        onlyGovernorOrAdmin
+    {
+        _setFloorBasisPoints(newFloorBasisPoints);
+    }
+
+    /// @notice sets the ceiling price in BP
+    function setOracleCeilingBasisPoints(uint256 newCeilingBasisPoints)
+        external
+        override
+        onlyGovernorOrAdmin
+    {
+        _setCeilingBasisPoints(newCeilingBasisPoints);
+    }
+
+    function isPriceValid() external view override returns (bool) {
+        return _validPrice(readOracle());
+    }
+
+    /// @notice helper function to set the ceiling in basis points
+    function _setCeilingBasisPoints(uint256 newCeilingBasisPoints) internal {
+        require(
+            newCeilingBasisPoints != 0,
+            "PegStabilityModule: invalid ceiling"
+        );
+        require(
+            Decimal
+                .ratio(
+                    newCeilingBasisPoints,
+                    Constants.BASIS_POINTS_GRANULARITY
+                )
+                .greaterThan(
+                    Decimal.ratio(floor, Constants.BASIS_POINTS_GRANULARITY)
+                ),
+            "PegStabilityModule: ceiling must be greater than floor"
+        );
+        uint256 oldCeiling = ceiling;
+        ceiling = newCeilingBasisPoints;
+
+        emit OracleCeilingUpdate(oldCeiling, ceiling);
+    }
+
+    /// @notice helper function to set the floor in basis points
+    function _setFloorBasisPoints(uint256 newFloorBasisPoints) internal {
+        require(newFloorBasisPoints != 0, "PegStabilityModule: invalid floor");
+        require(
+            Decimal
+                .ratio(newFloorBasisPoints, Constants.BASIS_POINTS_GRANULARITY)
+                .lessThan(
+                    Decimal.ratio(ceiling, Constants.BASIS_POINTS_GRANULARITY)
+                ),
+            "PegStabilityModule: floor must be less than ceiling"
+        );
+        uint256 oldFloor = floor;
+        floor = newFloorBasisPoints;
+
+        emit OracleFloorUpdate(oldFloor, floor);
+    }
+
+    /// @notice helper function to determine if price is within a valid range
+    function _validPrice(Decimal.D256 memory price)
+        internal
+        view
+        returns (bool valid)
+    {
```

Comment:
let's read in floor and ceiling into variables on the stack so that we can do a single SLOAD, instead of an additional warm SLOAD